// @cool-vue/ai/utils/ai.js
import { assign, isEmpty } from "lodash-es";
import { serviceHost } from "../config.js";
import axios from "axios";

export async function invokeFlow(label, params, loading) {
  const request = (params) => {
    return new Promise((resolve, reject) => {
      axios
        .post(
          serviceHost + "/open/flow/run/invoke",
          {
            params: {
              locale: params.locale,
              data: JSON.stringify(params.data),
            },
            label,
            stream: false,
          },
          {
            // responseType: "stream",
          }
        )
        .then((response) => {
          const { code, data, message } = response.data;

          if (code === 1000) {
            try {
              let content = data.result.res;

              if (content.includes("```json")) {
                content = content.split("```json")[1].split("```")[0];
              }

              const res = JSON.parse(content);
              const d = {};

              params.data.forEach((k, i) => {
                d[k] = res[i];
              });

              resolve(d);
            } catch (err) {
              console.log(data.result.res);

              reject({
                message: "Error, please try again",
              });
            }
          } else {
            reject({
              message,
            });
          }
        })
        .catch(reject);
    });
  };

  let data = {
    modules: {},
    plugins: {},
    uniappx: {},
    uni_modules: {},
  };

  const languages = params.languages.filter((e) => e !== "zh-cn");

  let task = [];

  for (let m in params.data) {
    const [type, module] = m.split(".");

    if (!data[type][module]) {
      data[type][module] = {
        "zh-cn": {},
      };

      params.data[m].forEach((e) => {
        data[type][module]["zh-cn"][e] = e.replace(/^page\./, "");
      });
    }

    languages.forEach((l) => {
      // 拆分
      const req = new Promise(async (resolve) => {
        const list = params.data[m];
        const data = {};

        // 每20个拆分一个
        const reqs = [];
        for (let i = 0; i < list.length; i += 20) {
          const arr = list.slice(i, i + 20);

          if (!isEmpty(arr)) {
            reqs.push(
              request({
                locale: l,
                data: arr,
              }).then((res) => assign(data, res))
            );
          }
        }

        await Promise.all(reqs);

        resolve(data);
      });

      task.push(
        req.then((res) => {
          data[type][module][l] = res;
          loading.succeed(`${module}[${l}]`);
        })
      );
    });
  }

  await Promise.all(task);

  return data;
}
